EIPS/eip-152.md

```diff
@@ -0,0 +1,140 @@
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+
+Besides being a useful cryptographic hash function and SHA3 finalist, BLAKE2b allows for efficient verification of the Equihash PoW used in Zcash, making a BTC Relay - style SPV client possible on Ethereum. A single verification of an Equihash PoW verification requires 512 iterations of the hash function, making verification of Zcash block headers prohibitively expensive if a Solidity implementation of BLAKE2b is used.
+
+The BLAKE2b algorithm is highly optimized for 64-bit CPUs, and is faster than MD5 on modern processors.
+
+Interoperability with Zcash could enable contracts like trustless atomic swaps between the chains, which could provide a much needed aspect of privacy to the very public Ethereum blockchain.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+We propose adding a precompiled contract at address `0x09` wrapping the [BLAkE2b `F` compression function](https://tools.ietf.org/html/rfc7693#section-3.2). Rather than the ABI-encoded call structure proposed in [#152](https://github.com/ethereum/EIPs/issues/152), this design better matches other precompiles and works well with a standard `staticcall`.
+
+The precompile requires 6 inputs, corresponding to the inputs specified in the [BLAKE2b RFC](https://tools.ietf.org/html/rfc7693#section-3.2)
+
+- `h` - the state vector
+- `m` - the message block vector
+- `t_0, t_1` - offset counters
+- `f` - the final block indicator flag
+- `rounds` - the number of rounds
+
+Loosely packing the data in equal 32-byte slots means an easier job for app developers, and also appears to be compatible with how other precompiles accept input data.
+
+```
+[ bytes32[0] | bytes32[1] ][ bytes32[2] | bytes32[3] | bytes32[4] | bytes32[5] ]
+[      64 bytes for h     ][                    128 bytes for m                ]
+
+[             bytes32[6]             ][             bytes32[7]             ]
+[ 24 bytes padding | 8 bytes for t_0 ][ 24 bytes padding | 8 bytes for t_1 ]
+
+[             bytes32[8]          ][              bytes32[9]              ]
+[ 31 bytes padding | 1 byte for f ][ 28 bytes padding | 4 bytes for rounds]
```

Comment:
Some data is marked as `padding`. Does the EIP place any restrictions on the padding? Does padding need to be zeroes? Should an error be thrown if padding is non-zero?

Reply:
An alternative no-padding solution has been proposed [here](https://github.com/ethereum/EIPs/pull/2129#discussion_r296652046).

Reply:
How would the new definition look? Spontaneously I think less padding is for the better...

Reply:
No padding at all - arguments will be ABI-encoded (packed together with `abi.encodePacked`):
```
[ 64 bytes for h ][ 128 bytes for m ][ 8 bytes for t_0][ 8 bytes for t_1 ][ 2 bytes for f ][ 4 bytes for rounds]
```

For some reason, `bool` `f` takes 2 bytes instead of one but it's how `abi.encodePacked` works.

Reply:
Are you sure it returns 2 byte for the bool? It should return 1.

Reply:
You are right @axic - it's 1 byte. I incorrectly split the resulting bytestring in my editor...

The encoded parameters look as follows:
```
[ 64 bytes for h ][ 128 bytes for m ][ 8 bytes for t_0][ 8 bytes for t_1 ][ 1 byte for f ][ 4 bytes for rounds]
```

Reply:
Updated this section with the proposed `abi.encodePacked` input packing.